src/server.ts

```diff
@@ -267,41 +256,58 @@ class TotalCount
+            const data = {
+                "package": dep.name.value,
+                "version": dep.version.value
+            };
+            dependencyMaping.set(`${dep.name.value}-${dep.version.value}`, dep);
+            payload.push(data);
+        } else {
+            aggregator.aggregate(dep);
+        } 
+    });
+    return {payload, dependencyMaping};
+};
+
 const regexVersion =  new RegExp(/^([a-zA-Z0-9]+\.)?([a-zA-Z0-9]+\.)?([a-zA-Z0-9]+\.)?([a-zA-Z0-9]+)$/);
-const sendDiagnostics = (ecosystem: string, uri: string, contents: string, collector: IDependencyCollector) => {
+const sendDiagnostics = async (ecosystem: string, uri: string, contents: string, collector: IDependencyCollector) => {
     connection.sendNotification('caNotification', {'data': caDefaultMsg});
-    collector.collect(contents).then((deps) => {
-        let diagnostics = [];
-        /* Aggregate asynchronous requests and send the diagnostics at once */
-        let aggregator = new Aggregator(deps, () => {
-            connection.sendNotification('caNotification', {'data': getCAmsg(deps, diagnostics, totalCount), 'diagCount' : diagnostics.length > 0? diagnostics.length : 0});
-            connection.sendDiagnostics({uri: uri, diagnostics: diagnostics});
-        });
-        
-        let totalCount = new TotalCount();
-
-        for (let dependency of deps) {
-            if(dependency.name.value && dependency.version.value && regexVersion.test(dependency.version.value.trim())) {
-                get_metadata(ecosystem, dependency.name.value, dependency.version.value).then((response) => {
-                    if (response != null) {
-                        let pipeline = new DiagnosticsPipeline(DiagnosticsEngines, dependency, config, diagnostics, uri);
-                        pipeline.run(response);
-                        for (const item of pipeline.items) {
-                            let secEng = item as SecurityEngine;
-                            totalCount.vulnerabilityCount += secEng.vulnerabilityCount;
-                            totalCount.advisoryCount += secEng.advisoryCount;
-                        }
-                    }
-                    aggregator.aggregate(dependency);
-                }).catch((err)=>{
-                    aggregator.aggregate(dependency);
-                    connection.console.log(`Error ${err} while ${dependency.name.value}:${dependency.version.value}`);
-                });
-            } else {
-                aggregator.aggregate(dependency);
-            }
-        }
+    const deps = await collector.collect(contents);
+    let diagnostics = [];
+    /* Aggregate asynchronous requests and send the diagnostics at once */
+    let aggregator = new Aggregator(deps, () => {
+        connection.sendNotification('caNotification', {'data': getCAmsg(deps, diagnostics, totalCount), 'diagCount' : diagnostics.length > 0? diagnostics.length : 0});
     });

```

Comment:
Use simplified reference count based aggregator which matches to request - response count to send completion notification.

Reply:
Also combine it with a timeout & fail after certain time.